### PR TITLE
Make IngestSegmentFirehoseFactory splittable for parallel ingestion

### PR DESCRIPTION
**Pull request by glasser**

### CODE REVIEW

**Comment by glasser** [comment]
I recognize that this is a PR without a matching issue, so it's more of a "proposal in the form of code" than a ready-to-merge pull request, and I will not feel like my time writing the code has been wasted if this change is unwanted. The PR currently has no tests; if the design of this feature is accepted, I'll write some.

**Comment by glasser** [comment]
Note: there's a comment in the PR about how `Granularity.getIterable` can return intervals which extend beyond its input interval.  Here are the uses of that function in non-test code in Druid:

- druid-indexing-hadoop, GranularityPathSpec: trims the output of that iterable
- druid-processing, IntervalChunkingQueryRunner: does some processing which I believe is equivalent to trimming the output of that iterable
- druid-processing, GroupByUniversalTimestamp: probably should be using granularity.bucketStart instead
- druid-processing, QueryableIndexStorageAdapter.CursorSequenceBuilder: appears to trim the output of that iterable (inside the `apply` method)
- druid-processing, IncrementalIndexStorageAdapter: most uses of the interval (in initializing timeStart and cursorIterable) trim it to the actualInterval, but the `time` field doesn't. Is that a bug?
- druid-server, UniformGranularitySpec: does not trim the intervals.  Because batch ingestion uses this field to throw away data not in the given explicit intervals, this means that the interval used to throw away data may be less precisely (larger) than the specified one.

This suggests that it would be helpful to have a `Granularity.getTrimmedIterable()` method which would allow most of the current callers of `Granularity.getIterable()` to be simplified. There are two calls in my list that wouldn't be able to use getTrimmedIterable, and I don't know Druid well enough to say whether that is due to bugs or is intentional.  If we do conclude that these are both bugs, I don't know the Druid compatibility rules well enough to know if it would be OK to just change the semantics of `Granularity.getIterable()` rather than add a second variant.

**Comment by jihoonson** [comment]
@glasser thanks for working on this! I have one comment.

As in other distributed systems like Hadoop or Spark, it's important to evenly distributed work in parallel indexing because the entire ingestion is completed when the longest task is finished. It means, each sub task is supposed to process almost same amount of data.
In this PR, it looks that input segments are split based on `taskGranularity`. But, in practice, it's possible that each time chunk has different amounts of data. What do you think about splitting based on the segment size, so that each task process a similar size of segments?

**Comment by glasser** [comment]
Good question. I was kind of imagining you would set taskGranularity equal to your output segmentGranularity so that each subtask would write one segment. You're right that things can get unbalanced though.

Are you imagining that the split implementation would query the segments metadata to learn all the segment sizes and the user would specify bytes per split?  Would we try to not divide any input segments but just chunk them together?

This seems like a reasonable option to desire but I kind of feel like people might still want to get started with the simpler "I know my peons can handle an hour of data, just split by hours" anyway... so implementing one of these options doesn't necessarily stop from implementing the other later.

**Comment by jihoonson** [comment]
> Are you imagining that the split implementation would query the segments metadata to learn all the segment sizes and the user would specify bytes per split? Would we try to not divide any input segments but just chunk them together?

Yes, this is exactly what I want. The task can ask the coordinator to get the segment metadata. Tasks use `CoordinatorClient` when they talk with the coordinator, so you may want to add a new method to it which calls `/druid/coordinator/v1/datasources/{dataSourceName}/intervals/{interval}?simple` (http://druid.io/docs/latest/operations/api-reference.html#datasources). Its result is like below:

```json
{
  "2019-02-11T23:00:00.000Z/2019-02-12T00:00:00.000Z": {
    "size": 21459255,
    "count": 2
  },
  "2019-02-11T22:00:00.000Z/2019-02-11T23:00:00.000Z": {
    "size": 24510542,
    "count": 2
  }
}
```

> This seems like a reasonable option to desire but I kind of feel like people might still want to get started with the simpler "I know my peons can handle an hour of data, just split by hours" anyway... so implementing one of these options doesn't necessarily stop from implementing the other later.

My feeling is that `maxInputSegmentBytesPerTask` is simpler than `taskGranularity` because , with `taskGranularity`, people should think about how many segments are in each time chunk and each task can handle it. However, for `maxInputSegmentBytesPerTask`, they can set it to whatever a task can handle. I think we can provide a default, so that people even don't have to think about it in most cases.

If `taskGranularity` is better than `maxInputSegmentBytesPerTask` in some cases, I'm fine with adding it. But, I don't think of anything. Do you have something in your mind?

**Comment by glasser** [comment]
I guess my logic is we run Kafka indexing service making one segment per task so handling one segment per task feels doable and simple :) But of course KIS can be configured with max bytes as well. I like the point that we can have a default max bytes here. 

What are you imagining this would do for single segments larger than the max bytes setting?

**Comment by glasser** [comment]
(er, single intervals, not single segments - I'm bad at terminology here.)

**Comment by jihoonson** [comment]
For an interval larger than the max bytes setting, I think each subTask should process a subset of segments, so that the large interval can be processed in parallel.

For a segment larger than the max bytes setting, probably it's fine to process with one subTask.

**Comment by glasser** [comment]
OK. And I guess it doesn't really matter how segments are assigned between tasks — all the locking and segment allocation works no matter how much overlap there is, right?

Since this will require the subtasks to understand how to have specific segments named rather than just intervals, perhaps the task itself should be extended to allow you to specify a list of segments as an alternative to an interval in the first place?

**Comment by jihoonson** [comment]
> OK. And I guess it doesn't really matter how segments are assigned between tasks — all the locking and segment allocation works no matter how much overlap there is, right?

Would you tell me more about what kind of overlap you're thinking? I think it's fine if each task processes a disjoint subset of input segments.

> Since this will require the subtasks to understand how to have specific segments named rather than just intervals, perhaps the task itself should be extended to allow you to specify a list of segments as an alternative to an interval in the first place?

I think we can extend `IngestSegmentFirehoseFactory` to accept a list of segments. In `IngestSegmentFirehoseFactory`, I think the `interval` should be empty if `segments` are specified. This `segments` should also be integrated with the `intervals` of `GranularitySpec`. If `intervals` are given, the task should read only the segments overlapping with the intervals.

**Comment by glasser** [comment]
> Would you tell me more about what kind of overlap you're thinking? I think it's fine if each task processes a disjoint subset of input segments.

I was just stressing about what happens about combining the results of processing two segments from a given time chunk on different subtasks, but I think the answer is that everything just works fine.

I have a few further questions about overlapping intervals.

(a) It looks like the `/druid/coordinator/v1/datasources/{dataSourceName}/intervals/{interval}` API only shows us segments that are fully contained inside the given interval. I think that's not what we want — we also want to include partially covered intervals (filtering the rows we read from them of course). Do you know if this happens to be accessible somewhere in the API? Would it be reasonable to add an `includeOverlapping` query parameter to that API which uses `theInterval::overlaps` instead of `theInterval::contains `for its filter.

Alternatively, it seems like we could just use the same `SegmentListUsedAction` that the `connect` method uses. Seems like this might be simpler and more clearly equivalent to the unsplit case.

In either case it looks like the client in question isn't immediately accessible. I think I would need to add an injection for CoordinatorClient directly into IngestSegmentFirehoseFactory, but it looks like several services use IndexingServiceFirehoseModule without binding CoordinatorClient. Would I want to add a binding for CoordinatorClient directly to IndexingServiceFirehoseModule?

Or if using SegmentListUsedAction I would have to make sure to call setToolbox in the supervisor task like it is currently called in the sub task.  That seems simpler than using CoordinatorClient.

(BTW I notice that when IndexTask calls setToolbox on its firehose factory, it knows how to recurse through CombiningFirehoseFactory, but the parallel sub task does not do that.  I think I could add a TaskToolboxConsumingFirehoseFactory interface implemented by IngestSegmentFirehoseFactory and CombiningFirehoseFactory to simplify this and make it more consistent.)

(b) I don't really know what happens when we have two segments with overlapping intervals and distinct versions.  Eg v1 @ 1:00-3:00, v2 @ 2:00-4:00.  Is the data from 1:00-2:00 in the first interval supposed to be visible, or is the entire segment overshadowed?  If it's the former, then specifically, when IngestSegmentFirehoseFactory calls 
```
      final List<TimelineObjectHolder<String, DataSegment>> timeLineSegments = VersionedIntervalTimeline
          .forSegments(usedSegments)
          .lookup(interval);
```

will the overshadowing still work if we're in a subtask where we're only passing the first segment in `usedSegments`, or does this calculation have enough context to understand that part of the segment should be ignored even though the segment that overshadows it is not in `usedSegments` in this particular subtask?

**Comment by jihoonson** [comment]
@glasser good questions!

For (a), thank you for reminding me about `SegmentListUsedAction`. I think it's the best option and you can set taskToolBox and use it. For `TaskToolboxConsumingFirehoseFactory`, I guess you want to extract `IndexTask.setFirehoseFactoryToolbox()` to `IngestSegmentFirehoseFactory`. Please go for it if you think it's simpler.

For (b), the first segment should be partially overshadowed. The part of v1 @ 1:00-2:00 should be available for input. `VersionedIntervalTimeline.lookup()` is responsible for doing this. If `usedSegments` are the segments of v1 @ 1:00-3:00 and v2 @ 2:00-4:00, `lookup()` would return 2 `TimelineObjectHolder` of v1 @ 1:00-2:00 and v2 @ 2:00-4:00. However, if `usedSegments` is v1 @ 1:00-3:00, `lookup()` just returns a `TimelineObjectHolder` for the same segment.

So, you may want to adjust `intervals` of the granularitySpec for sub tasks as well to prune out the overshadowed parts. I'm not sure how this problem is currently being handled though.

**Comment by jihoonson** [comment]
Oh, I found it. In `IngestSegmentFirehoseFactory`, it creates `WindowedStorageAdapter` which filters out data outside of the interval of `TimelineObjectHolder`. So, you don't have to worry about it.

**Comment by glasser** [comment]
But doesn't that holder come from the timeline, so it only works if the timeline was constructed from the full set of segments?

I think what this means is that `IngestSegmentFirehoseFactory.connect` needs to *always* call SegmentListAction/VersionedIntervalTimeline on the *full* original interval even in a subtask.  However it will only call fetchSegments on the selected segments, and it should skip elements of timelineSegments that aren't in the selected segments.

This implies that instead of `segments` being an alternate option for `interval`, `interval` always needs to be specified.

Alternatively, the split operation needs to provide both the full list of segments and the split-specific segment list to each split firehose factory.  (Or at least include an extra list of overlapping segments.)


(In other news, my TaskToolboxConsumingFirehoseFactory idea runs into trouble because CombiningFirehoseFactory is in druid-server which doesn't have access to druid-indexing-service's TaskToolbox type. I suppose the interface could declare `setTaskToolbox(Object)` and let IngestSegmentFirehoseFactory do a typecast?)

**Comment by jihoonson** [comment]
Is your concern that different sub tasks can process overlapping segments of different versions, so that both of them produce data of overlapping intervals which can cause the incorrect result? I think it's valid. 

To avoid this, we can add another restriction in split generation: each split should contain all overlapping segments. For example, suppose we have 3 segments of v1 @ 1:00-3:00, v2 @ 2:00-4:00, and v2 @ 4:00-6:00. Then, we need 2 sub tasks each of which processes (v1 @ 1:00-3:00 and v2 @ 2:00-4:00) and (v2 @ 4:00-6:00), respectively. Each sub tasks should be able to generate the valid timeline from the given subset of sub tasks.

I think this restriction is acceptable because this kind of overlapping won't be common. What do you think?

**Comment by glasser** [comment]
That sounds reasonable, with this caveat: you previously said:

> For an interval larger than the max bytes setting, I think each subTask should process a subset of segments, so that the large interval can be processed in parallel.

and I think what you meant by that is that if there are a bunch of segments with the same interval and version (different partition num) that it *would* be OK to split them up across sub tasks.

So I think the algorithm would be something like: list the segments for the whole interval as a timeline.  Select the first segment, and take the set of all segments that overlap it, transitively. If this set of segments has more than one interval, then all of those segments are constrained to go in the same subtask.  Otherwise, each of the segments in this set (all of which are for the same interval) may go in their own subtask.

We've now partitioned the full set of segments into subsets that have to go together.  We can then divide the whole list up into subtasks. This is the [https://en.wikipedia.org/wiki/Bin_packing_problem](https://en.wikipedia.org/wiki/Bin_packing_problem). Rather than try to optimally solve it or even use the first-fit problem, I would just use the greedy algorithm that goes down the list of subsets and greedily assigns them to subtasks, since this is more likely to get segments of the same interval onto the same subtask and thus lead to fewer output segments.

**Comment by jihoonson** [comment]
@glasser sounds nice! Are you planning to update your PR?

**Comment by glasser** [comment]
Yes, now that I think we have a concrete algorithm!

I did just pull out the observation I had above about the special-case setToolbox code in IndexTask not existing in ParallelIndexSubTask into https://github.com/apache/incubator-druid/pull/7063

**Comment by jihoonson** [comment]
Thanks! I'll take a look it too.

**Comment by glasser** [comment]
What do you think a good default value for maxInputSegmentBytesPerTask is? I was thinking of just copying what the Kafka indexing service uses for its default segment size limit, but that's a row count, not bytes.

**Comment by glasser** [comment]
> So I think the algorithm would be something like: list the segments for the whole interval as a timeline. Select the first segment, and take the set of all segments that overlap it, transitively. If this set of segments has more than one interval, then all of those segments are constrained to go in the same subtask. Otherwise, each of the segments in this set (all of which are for the same interval) may go in their own subtask.

OK, I think this was overcomplicating it. An easier way to solve the overlapping segment problem is just to make the task specify directly to the subtask a `List<WindowedSegment>`, where a WindowedSegment is a class that holds a segment ID (String) and a `List<Interval>`.  This can be read off directly from the full timeline when calculating splits.

The sub-task won't need to recalculate a timeline at all: it'll just set up some WindowedStorageAdapters to read the appropriate parts of the segments.

**Comment by glasser** [comment]
I'll work on this once #7089 is merged.
I think I can also change CompactionTask to use the new "tell IngestSegmentFirehoseFactory what segments to use directly" API.

**Comment by glasser** [comment]
@jihoonson OK, I have a basic implementation here. No new unit tests and haven't gotten all existing tests passing yet, but if you have a chance to take a basic look at the design that would be helpful! Will try to finish up the tests in the next couple days, but am curious if there are any major changes I'd need to make.

**Comment by jihoonson** [comment]
@glasser thanks! I'll take a look.

File: integration-tests/src/test/java/org/apache/druid/tests/indexer/AbstractITBatchIndexTest.java [hunk 1]
```diff
@@ -123,7 +123,11 @@ void doReindexTest(
       queryResponseTemplate = StringUtils.replace(
           queryResponseTemplate,
           "%%DATASOURCE%%",
-          fullBaseDatasourceName
+          // FIXME was the old value here a bug? Running queries against the original
```
Comment (glasser) [inline integration-tests/src/test/java/org/apache/druid/tests/indexer/AbstractITBatchIndexTest.java]: I pulled out this observation into its own PR: #7171

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java [hunk 2]
```diff
@@ -251,9 +273,169 @@ private long jitter(long input)
...
+    // FIXME This is doing a series of single HTTP calls for each segment. It could instead do a single
+    // GET /metadata/datasources/DATASOURCE/segments call to download all DataSegments. I'm erring on the
+    // side of avoiding single unboundedly-large calls, in favor of more smaller roundtrip --- after all,
+    // we're going to later fetch each of these segments in series, so the additional coordinator calls
+    // in series here don't feel that bad. And adding a new "get multiple segments by ID" HTTP call seems
+    // like overkill.
+    //
+    // Also: If the only use of WindowedSegments and the "segments" field on this FirehoseFactory is for internal
+    // use by parallel ingestion, then it might make sense to get rid of WindowedSegmentId entirely and just put
+    // WindowedSegments directly into this class's 'segments' parameter. Then we wouldn't even need to do this fetch
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: Another concern is that, Druid is using ZooKeeper for assigning tasks which means the overlord writes task specs in some ZooKeeper nodes to assign them. `SegmentIds` is safer than `DataSegments` because there's a limit for max bytes of data written in a node.

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegment.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A WindowedSegment represents a segment plus the list of intervals inside it which contribute to a timeline.
+ * If the list of intervals is null, the entire segment contributes to the timeline.
+ *
+ * This class is intended for in-memory use; WindowedSegmentId is better for serializing in specs.
+ */
+public class WindowedSegment
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegment.java]: The goal of this class looks similar to `TimelineObjectHolder`. `TimelineObjectHolder` is to represent each non-overshadowed logical segment backed by a real segment which might be partially overshadowed. `WindowedSegment` is opposite: It represents a real segment and what intervals are not overshadowed. Can we reuse `TimelineObjectHolder`?

File: docs/content/ingestion/firehose.md [hunk 4]
```diff
@@ -102,10 +103,12 @@ A sample ingest firehose spec is shown below -
 |--------|-----------|---------|
 |type|This should be "ingestSegment".|yes|
 |dataSource|A String defining the data source to fetch rows from, very similar to a table in a relational database|yes|
-|interval|A String representing ISO-8601 Interval. This defines the time range to fetch the data over.|yes|
+|interval|A String representing ISO-8601 Interval. This defines the time range to fetch the data over. You must specify exactly one of this and `segments`.|no|
+|segments|A list of JSON objects describing specific segments to include. Each object must have a `segmentId` field naming a segment to ingest, and may have an `intervals` field which is a list of ISO-8601 Interval strings telling what intervals on the segment to read. You must specify exactly one of this and `interval`.
```
[the remark this answers is not in the document]
Comment (glasser) [reply docs/content/ingestion/firehose.md]: OK, and document `interval` as required?

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegment.java (hunk 3 above)
Comment (glasser) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegment.java]: I think that's a good idea. I still want to have `WindowedSegmentId` (maybe with a different name) as `TimelineObjectHolder` isn't serializable, but I could reduce the delta of this PR by using `TimelineObjectHolder` instead of `WindowedSegment`.

The only trickiness is that the only code Druid has to construct `List<TimelineObjectHolder>` is `VersionedIntervalTimeline`, and [as described above](https://github.com/apache/incubator-druid/pull/7048#issuecomment-462904598) this isn't appropriate for a sub-task which only has a few segment IDs. So I have to construct it by hand, including making `PartitionChunk`s and such. But that's probably pretty straightforward.

File: docs/content/ingestion/firehose.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (glasser) [reply docs/content/ingestion/firehose.md]: Also, if the idea is to only use `segments` internally, I'll remove the support I had for splitting a task that has segments specified and only make the split work for interval tasks.

**Comment by glasser** [comment]
Thanks @jihoonson. It's much simpler using TimelineObjectHolder and considering `segmentIds` mode to be an internal detail.

It's probably worth another review pass now, though I still need to do unit tests. I'll wait to see what you think of #7171 before making the integration tests work.

**Comment by glasser** [comment]
Note that I'm holding off on cleaning up the integration tests here until https://github.com/apache/incubator-druid/pull/7171 is merged. I should write unit tests (eg of the splitting algorithm) though.

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegmentId.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import org.joda.time.Interval;
+
+import java.util.List;
+
+/**
+ * A WindowedSegment represents a segment plus the list of intervals inside it which contribute to a timeline.
+ * <p>
+ * This class is intended for serialization in specs.
+ */
+public class WindowedSegmentId
```
[the remark this answers is not in the document]
Comment (glasser) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegmentId.java]: The goal here is to ensure that we only send a given segment to one subtask, even if it has multiple windows. So it seemed reasonable to at least represent it in memory as an object that is "a segment with some windows". Does that seem right?

It also seemed nice to keep the representation that shows up in the task spec simpler (just segment ids, not full DataSegment) though maybe that is the wrong tradeoff. I was especially leaning towards that when the idea was that you might manually write a spec with segment IDs.

[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegmentId.java]: Oh I see. Sounds good.

**Comment by glasser** [comment]
I've added a unit test that I feel pretty good about, which takes advantage of the existing useful test cases in IngestSegmentFirehoseFactoryTimelineTest.

I have two questions about the impact on CompactionTask from this change that I had added as FIXME comments earlier. The last commit now removes those comments (so that the PR could be in theoretically-mergeable state) but I'd still appreciate feedback on those questions.

File: indexing-service/src/test/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java [hunk 6]
```diff
@@ -146,6 +161,36 @@ public void testSimple() throws Exception
...
+    long sum = 0;
+
+    for (InputSplit<List<WindowedSegmentId>> split : splits) {
+      final FiniteFirehoseFactory<InputRowParser, List<WindowedSegmentId>> splitFactory =
+          factory.withSplit(split);
+      try (final Firehose firehose = splitFactory.connect(ROW_PARSER, null)) {
+        while (firehose.hasMore()) {
+          final InputRow row = firehose.nextRow();
+          count++;
+          sum += row.getMetric(METRICS[0]).longValue();
```
Comment (glasser) [inline indexing-service/src/test/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java]: TeamCity complains that both method calls here may cause NPEs. It is just a duplicate of existing code with the same problem in testSimple, though.

File: indexing-service/src/test/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java [hunk 7]
```diff
@@ -286,13 +331,26 @@ private static InputRow ir(String timeString, long metricValue)
...
         }
+
+        @Override
+        public DataSegment getDatabaseSegmentDataSourceSegment(String dataSource, String segmentId)
+        {
+          return testCase.segments
+              .stream()
+              .filter(s -> s.getId().toString().equals(segmentId))
+              .findAny()
+              .get();  // throwing if not found is exactly what the real code does
```
Comment (glasser) [inline indexing-service/src/test/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactoryTimelineTest.java]: TeamCity doesn't want me to call `Optional.get()` without `ifPresent()` because it might throw, but I'm OK with it throwing. Not sure if I should change anything?

**Comment by glasser** [comment]
Most of the TeamCity errors are in untouched code. Two of them are in new code from me but I'm not sure what changes would be useful (see my comments).

**Comment by jihoonson** [comment]
It was probably because of the [changed Intellij version](https://lists.apache.org/thread.html/94182bbaac276c7cca2089be1493b42c0c31c010894887453a10ba94@%3Cdev.druid.apache.org%3E) which is reverted back now. I restarted TemCity.

**Comment by glasser** [comment]
Good, TeamCity passed this time.

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java [hunk 8]
```diff
@@ -251,6 +285,179 @@ private long jitter(long input)
...
+
+          // Now figure out if it goes in the current split or not.
+          final long segmentBytes = segment.getSize();
+          if (bytesInCurrentSplit + segmentBytes > maxInputSegmentBytesPerTask && !currentSplit.isEmpty()) {
+            // This segment won't fit in the current non-empty split, so this split is done.
+            newSplits.add(new InputSplit<>(currentSplit));
+            currentSplit = new ArrayList<>();
+            bytesInCurrentSplit = 0;
+          }
+          if (segmentBytes > maxInputSegmentBytesPerTask) {
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: Hmm, this might not be what we want. The idea of size-based split is to process roughly the same amount of data in each subTask. How ingestSegmentFirehose actually works is, 1) per segment, it reads the timestamp column and 2) reads other columns if the timestamp falls in the given interval (`QueryableIndexStorageAdapter.CursorSequenceBuilder`). 

But, this `if` clause compares the entire segment size with `maxInputSegmentBytesPerTask` even if the interval covers only a part of a segment. I guess we may need to estimate the segment size to be read roughly by computing based on the ratio of the window interval to the whole interval of a segment?

Comment (glasser) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: I'm happy to implement this today, but as a couple counterpoints before I do so:

- Size-based split does help keep subtasks equal, but it also has the purpose of bounding the amount of disk space needed for each task (because the current implementation downloads all segments at once at the beginning of the task), at least as long as all individual segments fall below the boundary.  It would be a shame if a single task ended up requiring an enormous amount of disk space (and download time) if it ended up being assigned a large number of segments which only contribute a small fraction of their data to the output segment.

- In practice I expect that most people's segments have uniform granularity (or at least are non-overlapping, eg older data having uniform day granularity and newer data having uniform hour granularity) and so the only time that windows will not be the full segment size is at the beginning and end of the total interval.  And even having misalignment there will be a mistake if you're using this to re-ingest data into itself with `appendToSegments: false` since it will drop data in the start and end segments that are outside of the interval.  So I think windowing is not going to be a super common use case in practice, and so doing a performance optimization that only matters for the windowing use case might not be worth the extra code complexity and more complicated documentation needed for `maxInputSegmentBytesPerTask`.  (Of course the overlapping segment use case needs to be supported for *correctness* because that's how the Druid data model works. But that doesn't mean it needs performance optimizations?)

Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: > In practice I expect that most people's segments have uniform granularity (or at least are non-overlapping, eg older data having uniform day granularity and newer data having uniform hour granularity) and so the only time that windows will not be the full segment size is at the beginning and end of the total interval. And even having misalignment there will be a mistake if you're using this to re-ingest data into itself with appendToSegments: false since it will drop data in the start and end segments that are outside of the interval. So I think windowing is not going to be a super common use case in practice, and so doing a performance optimization that only matters for the windowing use case might not be worth the extra code complexity and more complicated documentation needed for maxInputSegmentBytesPerTask. (Of course the overlapping segment use case needs to be supported for correctness because that's how the Druid data model works. But that doesn't mean it needs performance optimizations?)

If I understand correctly, windowing would happen depending on the value of `maxInputSegmentBytesPerTask` rather than segment granularity because split is created whenever `bytesInCurrentSplit + segmentBytes` is larger than `maxInputSegmentBytesPerTask`. Am I missing something?

> Size-based split does help keep subtasks equal, but it also has the purpose of bounding the amount of disk space needed for each task (because the current implementation downloads all segments at once at the beginning of the task), at least as long as all individual segments fall below the boundary. It would be a shame if a single task ended up requiring an enormous amount of disk space (and download time) if it ended up being assigned a large number of segments which only contribute a small fraction of their data to the output segment.

Good point. I think it would be an issue, but maybe it would be better to handle it in `IngestSegmentFirehoseFactory` like lazily downloading segments whenever needed (probably worth to combine with `PrefetchableTextFilesFirehoseFactory`?). But, this should be done in a separate PR and I'm fine with going as it is. Would you please add a comment about this concern?

Comment (glasser) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: > If I understand correctly, windowing would happen depending on the value of maxInputSegmentBytesPerTask rather than segment granularity because split is created whenever bytesInCurrentSplit + segmentBytes is larger than maxInputSegmentBytesPerTask. Am I missing something?

Individual input segments are never split up over multiple tasks. The only times that the window on a segment isn't that segment's full interval are when the segment goes outside of the overall task's boundaries or if part of the segment is overshadowed.  So my point in that paragraph was that in practice, the ratio you're describing will generally be 100%, and I'm not sure the added complexity (including making the meaning of `maxInputSegmentBytesPerTask` more complex) are worth it.

I will add a comment about lazily downloading segments. I suppose there's something positive about failing fast if segments can't be found, but that's probably not that important.

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java [hunk 9]
```diff
@@ -52,28 +56,38 @@
...
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+import java.util.stream.Stream;
 
-public class IngestSegmentFirehoseFactory implements FirehoseFactory<InputRowParser>
+public class IngestSegmentFirehoseFactory implements FiniteFirehoseFactory<InputRowParser, List<WindowedSegmentId>>
 {
   private static final EmittingLogger log = new EmittingLogger(IngestSegmentFirehoseFactory.class);
+  private static final long DEFAULT_MAX_INPUT_SEGMENT_BYTES_PER_TASK = 150 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (glasser) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: You know, I swear I did when I wrote this, but I can't remember now and I clearly didn't write it down. Do you have any suggestions?

The default `maxRowsPerSegment` for Kafka indexing seems like a reasonable place to look at to start, but then one has to think about how many bytes are in a typical row and how many segments we'd like each task to produce.  My default here is probably too low?

File: integration-tests/src/test/resources/indexer/wikipedia_parallel_ingest_segment_index_task.json [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        }
+      }
+    },
+    "ioConfig": {
+      "type": "index_parallel",
+      "firehose": {
+        "type": "ingestSegment",
+        "dataSource": "%%DATASOURCE%%",
+        "interval": "2013-08-31/2013-09-02",
+        "maxInputSegmentBytesPerTask": 1
```
[the remark this answers is not in the document]
Comment (glasser) [reply integration-tests/src/test/resources/indexer/wikipedia_parallel_ingest_segment_index_task.json]: Yes: that's what ensures that the test uses multiple subtasks.

File: indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java (hunk 8 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java]: Ok, it makes sense.
